TMGToolbox/src/assignment/preprocessing/remove_redundant_functions.py

```diff
@@ -0,0 +1,296 @@
+        
+    def run(self):
+        self.tool_run_msg = ""
+        self.TRACKER.reset()
+        
+        try:
+            self._Execute()
+        except Exception as e:
+            self.tool_run_msg = _m.PageBuilder.format_exception(
+                e, _traceback.format_exc(e))
+            raise
+        
+        self.tool_run_msg = _m.PageBuilder.format_info("Run complete.")
+    
+    def __call__(self, xtmf_ScenarioNumber, ExportFile):
+        
+        self.Scenario = _m.Modeller().emmebank.scenario(xtmf_ScenarioNumber)
+        self.ExportFile = ExportFile
+
+        if (self.Scenario is None):
+            raise Exception("Scenario %s was not found!" %xtmf_ScenarioNumber)
+        
+        try:
+            self._Execute()
+        except Exception as e:
+            msg = str(e) + "\n" + _traceback.format_exc(e)
+            raise Exception(msg)
+    
+    ##########################################################################################################    
+    
+    
+    def _Execute(self):
+        with _m.logbook_trace(name="{classname} v{version}".format(classname=(self.__class__.__name__), version=self.version),
+                                     attributes=self._GetAtts()):
+
+            # check functions in the Emme database
+            fcns = _M.emmebank.functions()
+            fdlist = []
+            ftlist = []
+            fplist = []
+            folist = []
+
+            for f in fcns:
+                f_type = str(f)[0:2]
+                if f_type == "fd":
+                    fdlist.append(str(f)[2:])
+                elif f_type == "ft":
+                    ftlist.append(str(f)[2:])
+                elif f_type == "fp":
+                    fplist.append(str(f)[2:])
+                else:
+                    folist.append(str(f)[2:])
+                    
+            fdlist_database = list(map(int, fdlist))
+            ftlist_database = list(map(int, ftlist))
+            fplist_database = list(map(int, fplist))
+            folist_database = list(map(int, folist))
+
+            print "There are %s Auto Volume Delay (fd) functions in the database." %len(fdlist_database)
+            print "There are %s Transit Time (ft) functions in the database." %len(ftlist_database)
+            print "There are %s Turn Penalty (fp) functions in the database." %len(fplist_database)
+            print "There are %s other types of functions in the database." %len(folist_database)
+
+
+            # check functions used in all scenarios/networks
+            scens = _M.emmebank.scenarios()
+            fdlist_net =[]
+            ftlist_net =[]
+            fplist_net =[]
+
+            for scen_id in scens:
+                lks = _M.emmebank.scenario(scen_id).get_network().links()
+                segs = _M.emmebank.scenario(scen_id).get_network().transit_segments()
+                trns = _M.emmebank.scenario(scen_id).get_network().turns()
```

Comment:
You should cache the network instead of loading it 3 times.

Reply:
Thank you, I will change that.